### DISCUSSION TITLE
Any solution to visit the board which configured with dhcp address?

### DISCUSSION THREAD

**Discussion by atline** [labgrid-project/labgrid#1096] [Q&A] [answered]
Hi, guys,

Despite of the serial console access, we may want to have a secondary connection to board.
I see we could use next in exporter:

```
NetworkService:
  address: 192.168.1.3
  username: root
```

But if I our boards were configured under a dhcp server, means the ip may changes. Then if labgrid has any solution to allow the second connection with ssh? You know, the board is there, when it's up, the upside software surely possible to get the ip of that board. If possible labgrid exporter can get the ip by itself?

The final aim is: user could use `labgrid-client -p place1 ssh` to directly access the already acquired board.

Thanks.

**Comment by Emantor** [comment, marked as answer]
This can be done, but requires your board to have a strategy at the moment. If you have a strategy, you can boot up the board within a state, login, get the current IP address and register the `NetworkService` yourself within the strategy. Thereafter `SSHDriver` connections should work. Since labgrid client will use an existing registered driver from the environment, this will work with the labgrid-client environment and state command line options i.e. `labgrid-client -c env.yaml -s ssh ssh` where `ssh` is a state within your `env.yaml` which boots up the board, gets the ip address and registers a `NetworkService`.

**Reply by atline** [reply]
Thanks @Emantor , I successfully fulfill my requirement with your suggestion. With a `MyStrategy` who dynamically register a `NetworkService` to `target`, it's now ok to use ssh.

But still, I have one concern about how to deploy that `MyStrategy`. Let me explain our scenario: we mainly use labgrid for so called boardfarm, without any test script there using pytest. Usually, different users have freedom to install labgrid-client on any of their machines.

So, problem comes here (maybe not problem, just `operation` issue):

1. Should I tell every user to `cd xxx/lib/python3.7/site-packages/labgrid/strategy`, and put `mystrategy.py` in that folder? Meanwhile open `__init__.py` to add `from .mystrategy import MyStrategy`? Or we have some trick there to ease this process?
2. How you suggest to maintain the `env.yaml` which used by `labgrid-client -c`? Previously, user just need to use `labgird-client -p imx8mm-evk-sh01 ssh`, while now they need a `env.yaml` on their computer? What I can think is to store these files in a git repo, and let user clone that before using labgrid-client, any better solution there something like labgrid-client directly download it from some specified url?

And, a side question not related to this topic but related to operation I think:

I saw you guys seldom apply tag to sourcecode, meanwhile just daily image on dockerhub. Do you expect followers to build the docker image for their selves based on the commit they wanted? 

I asked these questions because I want to know the best practice to use this project, so we possibly to add some ansible script to standardize the procedure inside the company. Thanks.

**Reply by atline** [reply]
I find `imports` in environment configuration could make the trick, this answers most of my concerns.

**Reply by Emantor** [reply]
> I saw you guys seldom apply tag to sourcecode, meanwhile just daily image on dockerhub. Do you expect followers to build the docker image for their selves based on the commit they wanted?
> 
> I asked these questions because I want to know the best practice to use this project, so we possibly to add some ansible script to standardize the procedure inside the company. Thanks.

The problem is that docker support was contributed and not really used by us. We should fix this for the next release since apparently people are relying on the docker images more than we thought. So yes, the plan is to have docker images based on the release tags and than also have docker build from latest master so they can optionally be used.

**Comment by jluebbe** [comment]
There is also support to collect information from a switch via SNMP (https://github.com/labgrid-project/labgrid/blob/master/labgrid/resource/ethernetport.py). Together with the ARP information from that network's router, you can get the IPs used on a specific port.

This is not integrated with the SSH driver, but a Strategy could easily collect that information and update the NetworkService with it.

**Reply by atline** [reply]
@jluebbe , thanks. I'm not sure of this feature, as some labs' private network is maintained by IT team, I even don't know where the ethernet hub is, and some sites' device even in company's public network which I even don't know which port it is, some cascade there I think.

So, I want to know does this solution has any limit? E.g. restrict to some type of ethernet hub or something else? What's the advantage compare to the solution "to get the ip from device itself"? If no limit, I will try to communicate with our IT team to know the detail.
